compose: Fix bug with compose banners on mobile width.

Ddharmani3 opened the pull request:
This commit lets compose banner to use more than one line on mobile widths(<576px) by wrapping .compose_banner_action_button and .banner_content in a flex .compose_banner_content_wrapper and making it column flex on width<576px shifting action button to a new line.

The close button is positioned top right on mobile widths.

CZO discussion: https://chat.zulip.org/#narrow/stream/101-design/topic/compose.20banners.20on.20narrow.20screens

Fixes #25847

<!-- Describe your pull request here.-->

Fixes: <!-- Issue link, or clear description.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

- Width>576px:
![image](https://github.com/zulip/zulip/assets/64723994/a24ceeb1-6bfa-43de-b608-6c52b6024cfb)

- Width = 500px:
![image](https://github.com/zulip/zulip/assets/64723994/700d7d2b-5602-4b8b-a426-6853aaa145a7)

- Width = 350px:
![image](https://github.com/zulip/zulip/assets/64723994/f96307ff-5d8b-49ec-8ac1-978eb04348eb)


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

terpimost commented:
mobile looks pretty cool, for Width>576px: I suggest to align the buttons to the top side instead of vertically middle, so it would be something like this:
<img width="447" alt="image" src="https://github.com/zulip/zulip/assets/1903309/a6dac256-f52c-41c5-8305-60f678a70c70">

prakhar1144 commented on web/styles/compose.css:
```diff
@@ -290,6 +299,7 @@
 
     .banner_content {
         flex-grow: 1;
```
I think we no longer need `flex-grow: 1` here. right?

prakhar1144 reviewed (commented):
Thanks, @Ddharmani3 for working on this.

I have left a few review comments and posted a screenshot below.

nitpick:
The commit message has a few grammatical mistakes (missing articles). (Mentioning it as the [docs suggests](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html#formatting-guidelines):
> Be sure to check your description for typos, spelling, and grammar mistakes; commit messages are important technical writing and English mistakes will distract reviewers from your ideas.

Thanks :)

prakhar1144 commented:
<details>
<summary>
<b>I think the action button should be in a column under the text at this width too.</b>
</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/176ddf69-47f6-4b02-8e78-9b375db6d6aa"></img>
</details>

Ddharmani3 commented:
Addresses the feedback above.
@prakhar1144 @terpimost This PR is ready for review now.
This is how it looks on 765 px:
![image](https://github.com/zulip/zulip/assets/64723994/7d91e30e-d207-46c0-a022-26b32f4a4946)

prakhar1144 commented on web/templates/compose_banner/compose_banner.hbs:
```diff
@@ -3,14 +3,16 @@
-    {{#if button_text}}
-    <button class="compose_banner_action_button{{#if hide_close_button}} right_edge{{/if}}" {{#if scheduling_message}}data-validation-trigger="schedule"{{/if}}>{{button_text}}</button>
-    {{/if}}
+    <div class="compose_banner_content_wrapper">
```
[CZO Discussion](https://chat.zulip.org/#narrow/stream/101-design/topic/compose.20banners.20on.20narrow.20screens/near/1588596)

prakhar1144 reviewed (commented):
We have a few merge conflicts to be resolved.

Reason: Check the [recent 3 commits.](https://github.com/zulip/zulip/commits/main/web/templates/compose_banner/compose_banner.hbs) -- June 10, 2023
From commit messages:
* rename `compose_banner_action_button` to `main-view-banner-action-button`.
* rename `compose_banner` to `main-view-banner`.

We'd need to make the changes accordingly in this PR.

Ddharmani3 commented:
@prakhar1144 rebased the PR. This PR is ready for review now.

terpimost commented:
Close button should be attached to the top right corner 
<img width="788" alt="image" src="https://github.com/zulip/zulip/assets/1903309/17f44a60-e871-4940-b4d1-bb1c6f5009d0">
I would still align text as it was to the left side.
Please confirm that this is a max size for such design.

Ddharmani3 commented:
> Please confirm that this is a max size for such design.
> 

@terpimost Yes, this is the max size (768px)

> I would still align text as it was to the left side.

I will update the PR to align it to the left.

> Close button should be attached to the top right corner

It is in the top right corner, but I guess you are pointing out that it is not looking in the corner; this is because the top padding is more than right padding(12px > 8px) hence it doesn't look align right?(I did this padding to make it center when design is only one line)

I will update the PR to make it look in the corner for width < 768 px.

What should be the vertical and horizontal padding in this case(both should be same to make it look align)? Should it be 8px, so horizontal padding remains consistent?

terpimost commented:
banner buttons should look like this 
<img width="708" alt="image" src="https://github.com/zulip/zulip/assets/1903309/c306b861-7b6b-403a-a882-87e3d8b0d4e8">
And I see it was like this here 
<img width="110" alt="image" src="https://github.com/zulip/zulip/assets/1903309/91867df8-9391-437c-96ac-af14ea2feb72">
But it has the bigger difference here 
<img width="64" alt="image" src="https://github.com/zulip/zulip/assets/1903309/5b07ab7f-63de-4768-bf8b-769f0e351991">

prakhar1144 commented on web/styles/compose.css:
```diff
@@ -277,19 +277,37 @@
 
-    .banner_content {
-        flex-grow: 1;
+    @media (width > $md_min) {
```
I think we should add an `=` here.
> @media (width >= $md_min)

Currently, at 768px, the close button is not properly aligned.
![Screenshot from 2023-06-14 22-50-23](https://github.com/zulip/zulip/assets/56781761/28a90d3a-43e9-4fbc-9858-5153d1f27ad8)

Ddharmani3 replied:
@prakhar1144, Updated the PR to tweak this, this PR is ready for review now.

prakhar1144 reviewed (commented):
Suggested a minor tweak, otherwise LGTM!

prakhar1144 reviewed (commented):
Looks good to me. -- Thanks!

timabbott commented:
@karlstolley can you review this? Visually it looks great to me in the screenshots, and the code seems mostly sane, so I was tempted to merge, so you should feel free to do so if you're happy, but figured it's worth your taking a look since you're working on the compose box anyway.

karlstolley commented:
Overall this looks good. I do think that the layout should better track the width of the message box. For example, at a 750px-wide viewport, we have a compose box of 736px wide (as measured by the dimensions of `#compose-content`). At this point, the button-below layout looks really kind of odd to me, as there is plenty of space for the two-column layout:

![banner-layout-750px](https://github.com/zulip/zulip/assets/170719/74c24e66-3645-4459-bc61-e32893cad065)

The difference is especially clear when you bump the viewport open to 770px:

![banner-layout-770px](https://github.com/zulip/zulip/assets/170719/4bf3aea3-4427-4416-88b0-12bceaf7004c)

There we've got a compose box that's just 483px wide, but it's getting the button-beside layout.

This is actually a compelling use-case for [container queries](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_container_queries), but support is [largely only in very recent browsers](https://developer.mozilla.org/en-US/docs/Web/CSS/@container#browser_compatibility).

I'm not sure what the best approach is here--to try to fiddle with and potentially introduce more media-query breakpoints, or to just leave the PR as-is and flag this as something to address as the work on the compose box continues.

What are your thoughts, @Ddharmani3?

karlstolley commented:
Good point. Additionally, with some clever use of flexbox and what's allowed to grow/shrink, it might be possible to have this fit the compose box well without need of container queries.

Happy to study this a bit and pair with you on it, @Ddharmani3, if you're interested.

Ddharmani3 commented:
@karlstolley @terpimost I have updated the PR to make the button take space as per the text.
Now it looks like this on 760px:
![image](https://github.com/zulip/zulip/assets/64723994/73fc6319-f1f9-42f9-8c3a-bddec8f38cb7)
![image](https://github.com/zulip/zulip/assets/64723994/cf3731ed-6a86-423c-a207-a49cfc23cac6)

Also fixed some other alignment issues.

But about @karlstolley's other concern that the 3-line text on 770px(>md_min) should have a button-below layout. I think it should be worked on but since this PR and issue #25847 are about fixing banners on mobile widths, we can look at it along with the compose box work as also suggested by you:
```
or to just leave the PR as-is and flag this as something to address as the work on the compose box continues.
```
@karlstolley, your thoughts on this?

terpimost commented:
I don't think we should stretch the button if it's on a separate line. Let it be as it is and if the width isn't enough word wrap the text inside the button


I don't suggest to have such a big vertical space between the button and the text:
<img width="649" alt="image" src="https://github.com/zulip/zulip/assets/1903309/5af462e7-0275-4404-847c-303c392dd889">

Weird effect here when there is a space under the text:
<img width="783" alt="image" src="https://github.com/zulip/zulip/assets/1903309/304cb45f-be7b-4505-bb49-cc2480ac6e8d">

karlstolley commented:
@Ddharmani3 are you open to pairing on this? That might be the more expedient way to proceed. Let me know if you've got 20-30 minutes to spare sometime.

alya commented:
@karlstolley perhaps you could finish this one out at this point?

karlstolley commented:
@alya, sure, I can do that.
